src/Core/System/Snippet/Service/TranslationMetadataLoader.php

```diff
@@ -0,0 +1,116 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\System\Snippet\Service;
+
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\GuzzleException;
+use League\Flysystem\Filesystem;
+use Psr\Http\Message\ResponseInterface;
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\System\Snippet\DataTransfer\Metadata\MetadataCollection;
+use Shopware\Core\System\Snippet\DataTransfer\Metadata\MetadataEntry;
+use Shopware\Core\System\Snippet\SnippetException;
+use Shopware\Core\System\Snippet\Struct\TranslationConfig;
+use Symfony\Component\Filesystem\Path;
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * @internal
+ *
+ * @phpstan-type DecodedMetadata array<string, array{locale: string, updatedAt: string, progress: int}>
+ */
+#[Package('discovery')]
+class TranslationMetadataLoader
+{
+    private const CROWDIN_METADATA_LOCK = 'crowdin-metadata.lock';
+
+    public function __construct(
+        private readonly TranslationConfig $config,
+        private readonly ClientInterface $client,
+        private readonly Filesystem $filesystem,
+    ) {
+    }
+
+    /**
+     * @param list<string> $locales
+     */
+    public function getUpdatedMetadata(array $locales): MetadataCollection
+    {
+        $path = $this->getPath();
+
+        $localMetadata = $this->getLocalMetadata($path);
+        $remoteMetadataArray = $this->fetchRemoteMetadataArray();
+
+        foreach ($locales as $locale) {
+            $remoteEntry = $remoteMetadataArray[$locale] ?? null;
+
+            if ($remoteEntry === null) {
+                continue;
+            }
+
+            $remoteResult = MetadataEntry::create($remoteEntry);
+            $localMetadata->addIfRequired($remoteResult);
+        }
+
+        return $localMetadata;
+    }
+
+    public function save(MetadataCollection $remoteMetadata): void
+    {
+        $path = $this->getPath();
+
+        $this->filesystem->write($path, $remoteMetadata->toJson());
+    }
+
+    protected function getPath(): string
+    {
+        return Path::join(TranslationLoader::TRANSLATION_DIR, self::CROWDIN_METADATA_LOCK);
+    }
+
+    private function downloadFile(): ResponseInterface
+    {
+        try {
+            return $this->client->request(Request::METHOD_GET, $this->config->metadataUrl);
+        } catch (GuzzleException $e) {
+            throw SnippetException::translationMetadataDownloadFailed($this->config->metadataUrl, $e);
+        }
+    }
+
+    /**
+     * @return DecodedMetadata
+     */
+    private function decode(string $content): array
+    {
+        $data = json_decode($content, true, \JSON_THROW_ON_ERROR);
+
+        return array_column($data, null, 'locale');
+    }
+
+    private function getLocalMetadata(string $path): MetadataCollection
+    {
+        if (!$this->filesystem->fileExists($path)) {
+            return new MetadataCollection();
+        }
+
+        $localMetadata = $this->filesystem->read($path);
```

Comment:
Shouldn't we use the flysystem here?

Reply:
it uses the flysystem, check the import: `use League\Flysystem\Filesystem;`